bf2fc6cc711aee1a0c2a/kas-fleet-manager
# feat: update capacity reporting in cloud providers region

## Description
We currently report capacity information in the /cloud_providers/{id}/region endpoint. For each region, there is a capacity object that includes information on the instance type and if the max capacity for that instance type in the region has been reached. 

e.g.
```
{
     "kind": "CloudRegion",
      "id": "eu-west-1",
      ...
      "capacity": [
         {
              "instance_type": "standard",
              "max_capacity_reached": true
         }
      ]
},
```

This will no longer work with multiple sizes i.e. there could be capacity for a size 'x1' but may not have capacity for 'x2'. Instead of having this boolean value represent this, we can return a list of sizes that can still be created on that region. 

The new response should look like the following:
```
{
     "kind": "CloudRegion",
      "id": "eu-west-1",
      ...
      "capacity": [
         {
              "instance_type": "standard",
              "max_capacity_reached": true // to be deprecated
              "available_sizes": [ "x1", "x2", ...] // it'll be an empty array if no capacity is left
         }
      ]
},
```

Example response: 
```
{
    "kind": "CloudRegionList",
    "page": 1,
    "size": 22,
    "total": 22,
    "items": [
          ...
          {
              "kind": "CloudRegion",
              "id": "sa-east-1",
              "display_name": "South America, São Paulo",
              "enabled": false,
              "supported_instance_types": [],
              "capacity": []
          },
          {
              "kind": "CloudRegion",
              "id": "us-east-1",
              "display_name": "US East, N. Virginia",
              "enabled": true,
              "supported_instance_types": [
                  "standard",
                  "eval"
              ],
              "capacity": [
                  {
                      "instance_type": "standard",
                      "max_capacity_reached": false,
                      "available_sizes": [
                          "x1",
                          "x2"
                      ]
                  },
                  {
                      "instance_type": "eval",
                      "max_capacity_reached": false,
                      "available_sizes": [
                          "x1"
                      ]
                  }
              ]
          },
          ...
     ]
}
```

## Verification Steps
- Review test changes and ensure they are passing
- [Optional] Manual verification:
  - Run KAS Fleet Manager
  - Create a dummy cluster on 'us-east-1'.
     - It should be in a `ready` state and supports `standard,eval` instance types.
     - Set the kafka instance limit to 0 
  - Call the cloud_provider/aws/regions endpoint.
  - Look for `us-east-1` in the list.
     - capacity should have 2 items: one for `standard` and one for `eval`. Each should have the following fields:
        - instance_type, max_capacity_reached and available_sizes. 
        - available_sizes should be an empty array and max_capacity_reached should be 'true'
   - Modify the dummy cluster and set its kafka instance limit to 1. Restart KAS Fleet Manager
     - Call the cloud_provider/aws/regions endpoint again. Look for `us-east-1` in the list. available_sizes for each instance types should now have "x1"
   - Create a `standard` Kafka instance.
     ```
     {
       "multi_az": true,
       "name": "test",
       "region": "us-east-1",
       "plan": "standard.x1"
     }
     ```
    - Call the cloud_provider/aws/regions endpoint again. Look for `us-east-1` in the list. available_sizes for each instance types should now be an empty array since capacity is now used up.
 

## Checklist (Definition of Done)
